pkg/modprovider/status_client.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package modprovider
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-jose/go-jose/v3/json"
+	"github.com/jackc/puddle/v2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+
+	"github.com/pulumi/pulumi-terraform-module/pkg/tfsandbox"
+)
+
+type resourceStatusClientPool interface {
+	Acquire(ctx context.Context, logger tfsandbox.Logger, address string) (resourceStatusClientLease, error)
+}
+
+type resourceStatusClientLease interface {
+	pulumirpc.ResourceStatusClient
+	Release()
+}
+
+type resourceStatusClientPoolImpl struct {
+	mutex sync.Mutex
+
+	// The pool of resource status clients indexed by address.
+	pool map[string]*puddle.Pool[resourceStatusHandle]
+}
+
+type resourceStatusHandle struct {
+	client pulumirpc.ResourceStatusClient
+	conn   *grpc.ClientConn
+}
+
+var _ resourceStatusClientPool = (*resourceStatusClientPoolImpl)(nil)
+
+func (p *resourceStatusClientPoolImpl) getOrCreatePool(
+	address string,
+) (*puddle.Pool[resourceStatusHandle], error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if p.pool == nil {
+		p.pool = make(map[string]*puddle.Pool[resourceStatusHandle])
+	}
+	pool, ok := p.pool[address]
+	if ok {
+		return pool, nil
+	}
+
+	pool, err := puddle.NewPool(&puddle.Config[resourceStatusHandle]{
+		MaxSize: 1,
```

Comment:
Is this a limitation on the engine side?

In general I think it would help to document this file, but I understand that some of this `view` functionality is still a work in progress.

Reply:
Not really a limitation - I've lifted this file to its own package and documented a bit. I think it's a little future-leaning but pretty complete. We need to talk to the gRPC status service. This code implements connection pooling so that we reuse the connection instead of reconnecting to the same address repeatedly. Currently it uses max=1 connection per address, but in practice there's only one address. I perhaps naively thought the code is mostly serial so we need at most 1, but come to think of it the same provider may be calling multiple instances.  I can bump this number to a small integer, say 16.